File: apps/web/app/(protected)/documents/feature/types/document.ts
// -------------------------
// Frontend return types
// -------------------------
export type SavedPage = {
  id: number;
  pageNumber: number;
  srcText: string;
  sectionId: number | null;
  sectionPageIndex: number | null;
  blocks: SavedPageBlock[];
};

export type SavedPageBlock = {
  id: number;
  blockIndex: number;
  blockType: "text" | "image";
  tag: string | null;
  text: string | null;
  documentImageId: number | null;
  alt: string | null;
};

export type LoadedSection = {
  id: number;
  title: string;
  depth: number;
  parentSectionId: number | null;
  orderIndex: number;
  firstPageNumber: number;
  lastPageNumber: number;
};

export type LoadedDocumentImage = {
  id: number;
  mediaType: string | null;
  byteLength: number;
};

export type LoadedDocument = {
  documentId: number;
  title: string;
  author: string | null;
  sourceKind: string;
  srcLang: string;
  tgtLang: string | null;
  pages: SavedPage[];
  sections: LoadedSection[];
  images: LoadedDocumentImage[];
};


// -------------------------
// Backend return types
// -------------------------
export type SavedPageDTO = {
  id: number;
  page_number: number;
  src_text: string;
  section_id: number | null;
  section_page_index: number | null;
  blocks: SavedPageBlockDTO[];
};

export type SavedPageBlockDTO = {
  id: number;
  block_index: number;
  block_type: "text" | "image";
  tag: string | null;
  text: string | null;
  document_image_id: number | null;
  alt: string | null;
};

export type LoadedSectionDTO = {
  id: number;
  title: string;
  depth: number;
  parent_section_id: number | null;
  order_index: number;
  first_page_number: number;
  last_page_number: number;
};

export type LoadedDocumentImageDTO = {
  id: number;
  media_type: string | null;
  byte_length: number;
};

export type LoadedDocumentDTO = {
  document_id: number;
  title: string;
  author: string | null;
  source_kind: string;
  src_lang: string;
  tgt_lang: string | null;
  pages: SavedPageDTO[];
  sections: LoadedSectionDTO[];
  images: LoadedDocumentImageDTO[];
};

File: services/api/src/api/parsing/pdf/document.py
from dataclasses import dataclass, field
from typing import Iterable, Literal, Optional

BlockType = Literal["text", "image"]


@dataclass(frozen=True)
class PdfBlock:
    type: BlockType
    block_index: int
    bbox: tuple[float, float, float, float] | None = None
    # -- text
    tag: Optional[str] = None
    text: Optional[str] = None
    # -- image
    image_key: Optional[str] = None
    alt: Optional[str] = None


@dataclass(frozen=True)
class PdfImage:
    image_key: str
    href: str
    media_type: Optional[str]
    byte_length: int
    data: Optional[bytes] = None


@dataclass(frozen=True)
class PdfSection:
    title: str
    key: str
    order_index: int
    first_page_number: int
    last_page_number: int
    depth: int = 0
    parent_title: Optional[str] = None
    parent_key: Optional[str] = None


@dataclass
class PdfPage:
    page_number: int
    section_key: str
    section_title: str
    section_page_index: int
    blocks: tuple[PdfBlock, ...]

    def get_page_text(self) -> str:
        """
        Builds full page text from page blocks
        """
        text_blocks = [block.text for block in self.blocks if block.type == "text" and block.text]
        return "\n\n".join(text_blocks).strip()


@dataclass
class PdfDocument:
    title: Optional[str]
    author: Optional[str]
    language: Optional[str]
    cover_bytes: bytes | None = None
    cover_media_type: str | None = None
    sections: list[PdfSection] = field(default_factory=list)
    pages: list[PdfPage] = field(default_factory=list)
    images: dict[str, PdfImage] = field(default_factory=dict)

    def iter_content_pages(self) -> Iterable[PdfPage]:
        for page in self.pages:
            if not page.blocks:
                continue
            yield page

    def iter_content_blocks(self) -> Iterable[PdfBlock]:
        for page in self.iter_content_pages():
            for block in page.blocks:
                yield block

    def get_full_text(self) -> str:
        page_texts = [page.get_page_text() for page in self.iter_content_pages()]
        return "\n\n".join(text for text in page_texts if text).strip()
